src/elf.rs

```diff
@@ -370,27 +370,248 @@ impl<C: ContextObject> Executable<C> {
         })
     }
 
-    /// Fully loads an ELF, including validation and relocation
+    /// Fully loads an ELF
     pub fn load(bytes: &[u8], loader: Arc<BuiltinProgram<C>>) -> Result<Self, ElfError> {
-        // The new parser creates references from the input byte slice, so
-        // it must be properly aligned. We assume that HOST_ALIGN is a
-        // multiple of the ELF "natural" alignment. See test_load_unaligned.
-        let aligned;
-        let bytes = if is_memory_aligned(bytes.as_ptr() as usize, HOST_ALIGN) {
-            bytes
+        let e_flags = LittleEndian::read_u32(bytes.get(48..52).ok_or(ElfParserError::OutOfBounds)?);
+        let config = loader.get_config();
+        let sbpf_version = if config.enabled_sbpf_versions.end() == &SBPFVersion::V1 {
+            if e_flags == EF_SBPF_V2 {
+                SBPFVersion::V2
+            } else {
+                SBPFVersion::V1
+            }
         } else {
-            aligned = AlignedMemory::<{ HOST_ALIGN }>::from_slice(bytes);
-            aligned.as_slice()
+            match e_flags {
+                0 => SBPFVersion::V1,
+                EF_SBPF_V2 => SBPFVersion::V2,
+                _ => SBPFVersion::Reserved,
+            }
         };
-        Self::load_with_parser(&Elf64::parse(bytes)?, bytes, loader)
+        if !config.enabled_sbpf_versions.contains(&sbpf_version) {
+            return Err(ElfError::UnsupportedSBPFVersion);
+        }
+
+        let mut executable = if sbpf_version.enable_stricter_elf_headers() {
+            Self::load_with_strict_parser(bytes, loader)?
+        } else {
+            Self::load_with_lenient_parser(bytes, loader)?
+        };
+        executable.sbpf_version = sbpf_version;
+        Ok(executable)
     }
 
-    fn load_with_parser(
-        elf: &Elf64,
+    /// Loads an ELF without relocation
+    fn load_with_strict_parser(
+        bytes: &[u8],
+        loader: Arc<BuiltinProgram<C>>,
+    ) -> Result<Self, ElfParserError> {
+        use crate::elf_parser::{
+            consts::{
+                ELFMAG, EV_CURRENT, PF_R, PF_W, PF_X, PT_GNU_STACK, PT_LOAD, SHN_UNDEF, STT_FUNC,
+            },
+            types::{Elf64Ehdr, Elf64Shdr, Elf64Sym},
+        };
+
+        let aligned_memory = AlignedMemory::<{ HOST_ALIGN }>::from_slice(bytes);
```

Comment:
This is still copying the ELF, ins't it?

Reply:
Yes, one is still required because of loader-v3 misalignment.

Reply:
Is it solved with loader-v4?

Reply:
yes, I hope so, unless there is another source of misalignment I am unaware of.